react-backend/client/src/cb-admin/pages/Activities.js

```diff
@@ -0,0 +1,267 @@
+import { Activities } from '../../api';
+import ActivityLabel from '../components/ActivityLabel';
+
+
+const Nav = styled.nav`
+  display: flex;
+  align-items: center;
+  justify-content: space-between;
+  flex: 1;
+`;
+
+const FlexItem = styled.div`
+  flex: ${props => props.flex || '1'};
+`;
+
+const SubmitButton = PrimaryButton.extend`
+  margin-left: 2em;
+  height: 3em;
+`;
+
+const Table = styled.table`
+  background: transparent;
+  width: 100%;
+  padding: 2em;
+`;
+const TableHead = styled.thead``;
+const TableBody = styled.tbody``;
+const TableRow = styled.tr`
+  height: 3em;
+`;
+const TableCell = styled.td`
+  text-align: ${props => (props.center ? 'center' : 'left')};
+`;
+const TableHeader = styled.th``;
+
+
+const categories = [
+  { key: '0', value: '' },
+  { key: '1', value: 'Sports' },
+  { key: '2', value: 'Arts, Craft, and Music' },
+  { key: '3', value: 'Physical health and wellbeing' },
+  { key: '4', value: 'Socialising' },
+  { key: '5', value: 'Adult skills building' },
+  { key: '6', value: 'Education support' },
+  { key: '7', value: 'Employment support' },
+  { key: '8', value: 'Business support' },
+  { key: '9', value: 'Care service' },
+  { key: '10', value: 'Mental health support' },
+  { key: '11', value: 'Housing support' },
+  { key: '12', value: 'Work space' },
+  { key: '13', value: 'Food' },
+  { key: '14', value: 'Outdoor work and gardening' },
+  { key: '15', value: 'Local products' },
+  { key: '16', value: 'Environment and conservation work' },
+  { key: '17', value: 'Transport' },
+];
+
+const keyMap = {
+  name: 'Activity',
+  monday: 'Mon',
+  tuesday: 'Tue',
+  wednesday: 'Wed',
+  thursday: 'Thu',
+  friday: 'Fri',
+  saturday: 'Sat',
+  sunday: 'Sun',
+};
+const colToState = invertObj(keyMap);
+const columns = Object.values(keyMap);
+
+
+export default class ActivitiesPage extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      activities: {
+        items: {},
+        order: [],
+      },
+      form: {},
+      errors: {},
+    };
+  }
+
+  componentDidMount() {
+    Activities.get(this.props.auth)
+      .then((res) => {
+        const activities = res.data.result;
+
+        const order = activities.map(activity => activity.id);
+        const items = activities.reduce((acc, activity) => {
+          acc[activity.id] = activity;
+          return acc;
+        }, {});
+        this.setState(state => ({ ...state, activities: { items, order } }));
+
+        this.props.updateAdminToken(res.headers.authorization);
+      })
+      .catch((error) => {
+        switch (error.status) {
+          case 401:
+            return this.props.history.push('/admin/login');
+
+          case 500:
+            return this.props.history.push('/internalServerError');
+
+          default:
+            return this.props.history.push('/admin/login');
+        }
+      });
+  }
+
+  onChange = e =>
+    this.setState(assocPath(['form', e.target.name], e.target.value))
```

Comment:
In future, I reckon it's more readable not using `assocPath` in `setState`

Reply:
Depends how deeply the state is nested. Object spreading can get very unreadable.